lib/puppetserver/ca/action/prune.rb

```diff
@@ -0,0 +1,130 @@
+  Prune the list of revoked certificates of any duplication within it.
+
+Options:
+BANNER
+
+        def initialize(logger)
+          @logger = logger
+        end
+
+        def self.parser(parsed = {})
+          OptionParser.new do |opts|
+            opts.banner = BANNER
+            opts.on('--help', 'Display this command-specific help output') do |help|
+              parsed['help'] = true
+            end
+            opts.on('--config CONF', 'Path to the puppet.conf file on disk') do |conf|
+              parsed['config'] = conf
+            end
+          end
+        end
+
+        def run(inputs)
+          config_path = inputs['config']
+
+          # Validate the config path.
+          if config_path
+            errors = FileSystem.validate_file_paths(config_path)
+            return 1 if Errors.handle_with_usage(@logger, errors)
+          end
+
+          # Validate puppet config setting.
+          settings_overrides = {}
+          puppet = Config::Puppet.new(config_path)
+          puppet.load(settings_overrides)
+          return 1 if Errors.handle_with_usage(@logger, puppet.errors)
+
+          # Validate that we are offline
+          return 1 if HttpClient.check_server_online(puppet.settings, @logger)
+
+          # Getting the CRL(s)
+          loader = X509Loader.new(puppet.settings[:cacert], puppet.settings[:cakey], puppet.settings[:cacrl])
+
+          # Prune the CRL(s)
+          crl_list = loader.crls
+          prune_CRLs(crl_list)
+          write_crl_to_file(crl_list, puppet.settings[:cacrl])
+
+          @logger.inform("Finished pruning your local CRLs")
+          return 0
+        end
+
+        def prune_CRLs(crl_list)
+          crl_list.each do |list|
+            existed_serial_number = Set.new()
+            revoked_list = list.revoked
+            revoked_list.delete_if do |revoked|
+              if existed_serial_number.add?(revoked.serial)
+                false
+              else
+                debug_output(list, revoked) unless @logger.level < 4
+                true
+              end
+            end
+            list.revoked=(revoked_list)
+            puts "Finished pruning CRL of #{list.issuer.to_s(OpenSSL::X509::Name::RFC2253)}\n"
+          end
+        end
+
+        def write_crl_to_file(crl_to_write, file_path)
+          begin
+            file = File.open(file_path, "w")
+            crl_to_write.each do |crl|
+              file.write(crl.to_pem)
+            end
+          rescue IOError => e
+            @logger.err("Error when write out to file on disk")
+            @logger.err("Error: " + e.message)
+          ensure
+            file.close unless file.nil?
+          end
+        end
+
+        # Output for debug
+        #  - Name of issuer
+        #  - Reason for revocation
+        def debug_output(curr_crl, curr_revoked_cert)
+          issuer_name = curr_crl.issuer.to_s(OpenSSL::X509::Name::RFC2253)
+          extensions = curr_revoked_cert.extensions.join(', ')
```

Comment:
Honestly I think the serial number is probably the only really interesting thing that needs to be logged at debug (certname would be nice, but that's hard/slow/unreliable)

Reply:
I think the debug would be more like: "Removing duplicate of #{serial}" or "Removing duplicate of #{serial} from #{revocation_time}"

Reply:
If we did include the issuer, it'd be nice to get the issuer name at the top of the calling loop and reuse if for every deduped revocation.

Reply:
Ooh good idea. That's one way we can include the issuer name without it being slow.